packages/flutter_tools/lib/src/windows/visual_studio.dart

```diff
@@ -416,17 +419,32 @@ class VswhereDetails {
 
     return VswhereDetails(
       meetsRequirements: meetsRequirements,
-      installationPath: details['installationPath'] as String?,
-      displayName: details['displayName'] as String?,
-      fullVersion: details['installationVersion'] as String?,
+      installationPath: _validateString(details['installationPath'] as String?),
+      displayName: _validateString(details['displayName'] as String?),
+      fullVersion: _validateString(details['installationVersion'] as String?),
       isComplete: details['isComplete'] as bool?,
       isLaunchable: details['isLaunchable'] as bool?,
       isRebootRequired: details['isRebootRequired'] as bool?,
       isPrerelease: details['isPrerelease'] as bool?,
-      catalogDisplayVersion: catalog == null ? null : catalog['productDisplayVersion'] as String?,
+      catalogDisplayVersion: catalog == null ? null : _validateString(catalog['productDisplayVersion'] as String?),
```

Comment:
I would suggest not validating this or display name; all they are used for is showing to the user. Everything will still work even if they are invalid, so there's no reason to exit the tool (which will, among other things, prevent building for Windows).

Reply:
Hello, thanks for the feedback! I'm conflicted on this suggestion. I agree there's value in making the tool resilient to crashes, thereby also avoiding future bug fixes that need to be cherrypicked. That said, how do we ensure that `productDisplayVersion` is only used for display purposes going forward? Imagine in the future we add logic on this value, we may have subtle bugs due to replacement chars "leaking in". These issues would likely be harder to debug than simply removing an unnecessary check. Personally I'd prefer to be more strict until we have no other choice but to introduce more replacement char "holes".

Reply:
> That said, how do we ensure that `productDisplayVersion` is only used for display purposes going forward?

I'm not sure I understand the concern. For what purpose would we use freeform text that we don't control, other than display?

> Imagine in the future we add logic on this value

Could you describe an example of logic we could add on freeform, *localized* text that we have absolutely no control over and could be changed by Microsoft at literally any time?

> we may have subtle bugs (perhaps even crashes!) due to replacement chars "leaking in".

I'm not clear on how we could crash in that scenario, whereas what you have here guarantees—explicitly—that we crash.

Reply:
> We already have logic that uses this freeform text that we don't control. See this CMake workaround:
> 
> https://github.com/flutter/flutter/blob/ec20ea80ad98007bd022648168e7b3cc38b2e665/packages/flutter_tools/lib/src/windows/build_windows.dart#L73-L75

That's arguably a mistake; I really should have added a (non-display) `version` based on parsing `productSemanticVersion`, exposed that as a `Version`, and used that for the workaround.

(I say "arguably" because that workaround was, at the time it landed, already only intended to apply to one exact, historical version, and not something like a version range. So if the next version of VS had switched the "display version" to something arbitrary it wouldn't have mattered. That's why this is doing string comparison and not parsing it as a `Version`.)

> It's difficult to ensure no one ever uses the display version for logic, especially as layers of abstractions get added over time.

It's hard to ensure it, true, but the code would be wrong anyway. Taking the example above: if someone had written code like that but actually tried to parse it like a `Version` and do logic on it, it would be broken if Microsoft started making the display versions, say, alphabetical dog breeds instead of numbers, which they could absolutely do. Even if the dog breeds used correct UTF-8.

> > > we may have subtle bugs (perhaps even crashes!) due to replacement chars "leaking in".
> > 
> > 
> > I'm not clear on how we could crash in that scenario, whereas what you have here guarantees—explicitly—that we crash.
> 
> One example could be another `throwToolExit` if a string unexpectedly contains replacement chars.

I don't follow. I'm saying you should remove the code that will cause this to (unnecessarily) crash, 100% of the time, if there are replacement characters in the display strings, and the argument is that removing that code would be bad because at some later point someone might do exactly that same thing and the resulting crash would be bad?

Reply:
That seems reasonable. I imagine the odds of these fields including bad UTF-8 (particularly in the future, fixed version of vswhere) is relatively low, but you're right that these are display fields that, unlike filesystem paths, we don't actually rely on in any consequential way other than display (and the one special-case check for the bad cmake release), where a stray replacement character isn't the end of the world.

Reply:
Agreed. Stuart and I had a chat and landed on the same decision. I'll update this and add comments to make this clear

Reply:
Done! Please let me know if you have additional feedback 😄